### PR TITLE
split CI checks into separate jobs; closes #3

### LINKED ISSUE #3: split tests into separate jobs
Certain packages in this repo want various versions of things during testing (see #2).  In order to accommodate this without running way more tests than we need to on each build, we should split the jobs up in our GHA config and support this via npm scripts or some other means.

### PR DESCRIPTION
**Pull request by boneskull**
This splits the checks into two buckets: one for tests in `ios-simulator` which require additional prerequisites in the environment, and the other for everything else (which does not require a special environment other than macOS).

- for performance, use lockfiles and npm caching in CI
- add brew caching for `ios-simulator`
- remove `.azure-pipelines.yml` (for now)
- fix relative reference in `devDependencies`
- add `test:*` npm scripts
- add `os` and `publishConfig` properties to base `package.json`; sync across packages
- add `scripts/test-ios-simulator.sh` which configures and runs checks against `ios-simulator` in CI

Further, changeset 6163fc2 disables the `import/no-unresolved` rule.  This rule is problematic, because it depends upon the packages having been built (transpiled), which is orthogonal to code linting.

Resolves #3.

### CODE REVIEW

**Comment by boneskull** [comment]
@mykola-mokhnach As per your request, `node-simctl` has been removed from this monorepo.  I've translated the tests as described in the existing azure templates to just use GitHub's CI.  I know you have reasons for wanting to retain the Azure tests--I just wanted to give running it all in GHA a shot, and I'm happy to restore the Azure builds if this proves insufficient.

**Comment by boneskull** [comment]
I have "formally" asked GitHub for access to the Big Sur environment, so we'll see what they say.

**Comment by boneskull** [comment]
These tests are running very slowly...

File: .eslintrc [hunk 1]
```diff
@@ -1,5 +1,6 @@
 {
   "extends": "@appium/eslint-config-appium",
+  "rules": {"import/no-unresolved": "off"},
```
Comment (mykola-mokhnach) [inline .eslintrc:3]: is this disabled temporarily?

File: .github/workflows/node.js.yml [hunk 2]
```diff
@@ -6,19 +6,58 @@ name: Node.js CI
...
+        with:
+          useRollingCache: true
+      - run: npm run test:common
 
+  ios-simulator:
+    name: ios-simulator checks (Node.js v${{ matrix.node-version }} / iOS v${{ matrix.ios-version }} / Xcode v${{ matrix.xcode-version }} / ${{ matrix.device-name }})
+    runs-on: macos-latest
+    strategy:
+      matrix:
+        node-version: [12.x, 14.x, 16.x]
```
Comment (mykola-mokhnach) [inline .github/workflows/node.js.yml:27]: I'd rather prefer to use the recent node version and different xcode versions in the matrix here.

File: package.json [hunk 3]
```diff
@@ -79,5 +87,11 @@
   },
   "engines": {
     "node": ">=12"
+  },
+  "os": [
```
Comment (mykola-mokhnach) [inline package.json:91]: I wouldn't do that

Comment (boneskull) [reply package.json:91]: ..?

**Comment by boneskull** [comment]
[Here's an example of why we should disable the ESLint rule.](https://github.com/appium/appium-ios/pull/5/checks?check_run_id=2916058157#step:5:160)  If package _x_ depends on _y_, and _x_ is linted first, this happens.

File: .eslintrc (hunk 1 above)
Comment (boneskull) [reply .eslintrc:3]: no, see https://github.com/appium/appium-ios/pull/4#issuecomment-868729225

This rule does not work well for a monorepo combined with how we build and publish packages

**Comment by boneskull** [comment]
@mykola-mokhnach I've added more checks for Xcode versions running against the default Node.js.

**Review by jlipps — APPROVED** [review]
seems reasonable!

File: .github/workflows/node.js.yml [hunk 4]
```diff
@@ -6,19 +6,64 @@ name: Node.js CI
...
+        run: |
+          brew tap wix/brew
+          brew update --preinstall
+          cat "$(brew --repository)/Library/Taps/wix/homebrew-brew/applesimutils.rb" > .github/brew-formulae
+      - name: Install Homebrew dependencies
+        env:
+          HOMEBREW_NO_AUTO_UPDATE: 1
+        run: brew install applesimutils
+      - name: Tests (Node.js v${{ matrix.node-version }} / iOS v${{ matrix.ios-version }} / Xcode v${{ matrix.xcode-version }} / ${{ matrix.device-name }})
+        run: scripts/test-ios-simulator.sh
```
Comment (mykola-mokhnach) [inline .github/workflows/node.js.yml:65]: I'm not quite sure if we want to combine unit and integration tests. I usually split them into stages, because it does not make sense to run e2e stage if unit tests fail. Also, for unit tests it makes more sense to try different node version and single xcode version while for e2e it's vice-versa (latest node and multiple Xcodes)

File: package.json (hunk 3 above)
Comment (boneskull) [reply package.json:91]: I've added `linux` to the `os` list

File: .github/workflows/node.js.yml (hunk 4 above)
Comment (boneskull) [reply .github/workflows/node.js.yml:65]: OK, I will split them up.

File: .github/workflows/node.js.yml [hunk 5]
```diff
@@ -6,19 +6,107 @@ name: Node.js CI
...
+          restore-keys: brew-
+      - name: Update Homebrew
+        run: |
+          brew tap wix/brew
+          brew update --preinstall
+          cat "$(brew --repository)/Library/Taps/wix/homebrew-brew/applesimutils.rb" > .github/brew-formulae
+      - name: Install Homebrew dependencies
+        env:
+          HOMEBREW_NO_AUTO_UPDATE: 1
+        run: brew install applesimutils
```
Comment (mykola-mokhnach) [inline .github/workflows/node.js.yml:57]: do we really need this all stuff to run unit tests?

File: .github/workflows/node.js.yml [hunk 6]
```diff
@@ -6,19 +6,107 @@ name: Node.js CI
...
 
 jobs:
-  build:
-
+  common:
+    name: Common tests (Node.js v${{ matrix.node-version }})
     runs-on: macos-latest
+    strategy:
+      matrix:
+        node-version: [12.x, 14.x, 16.x]
```
Comment (mykola-mokhnach) [inline .github/workflows/node.js.yml:14]: would it be possible to move these versions into common variables section, we we could bump them for all affected jobs when necessary?

Comment (mykola-mokhnach) [reply .github/workflows/node.js.yml:14]: same about xcode/device versions

File: .github/workflows/node.js.yml (hunk 5 above)
Comment (boneskull) [reply .github/workflows/node.js.yml:57]: probably not, I'll remove them and see

File: .github/workflows/node.js.yml (hunk 6 above)
Comment (boneskull) [reply .github/workflows/node.js.yml:14]: not sure if github supports this, or failing that, if it supports yaml aliases.  I'll see what I can do

Comment (boneskull) [reply .github/workflows/node.js.yml]: this is currently not possible:

- The [`defaults`](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#defaults) prop should support this sort of thing, but it only currently supports providing defaults for the `run` prop of a job.  The `strategy` prop of a job (which includes the `matrix`) [does not have access](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#context-availability) to any user-configurable Contexts.  I tried to create a JSON environment variable (via the top-level `env` prop) and use the `fromJSON()` function, but again, `strategy.matrix` cannot access the `env` Context.
- YAML anchors are not supported in workflow files, so I can't just use an alias.  Tried it!

I can't think of any other way to do this at the moment.

**Comment by boneskull** [comment]
Note that the `ios-simulator` E2E tests now _depend on_ the successful execution of the `ios-simulator` unit tests.

The breakdown of the build matrix is:

1.  All checks (unit, e2e) _other_ than `ios-simulator` run against supported Node.js versions
2. The `ios-simulator` unit tests run 
3. Assuming 2. completes successfully, `ios-simulator`'s e2e tests will run against the default Node.js version (I think it's v14 atm)

Steps 1 and 2 happen concurrently up to whatever GHA's concurrency limit is.

File: .github/workflows/node.js.yml (hunk 5 above)
Comment (boneskull) [reply .github/workflows/node.js.yml]: It doesn't appear they are needed.  removed

**Comment by boneskull** [comment]
This should be ready to merge pending green build.

File: .github/workflows/node.js.yml (hunk 6 above)
Comment (mykola-mokhnach) [reply .github/workflows/node.js.yml:14]: yep, Azure seems to be more advanced from that perspective

**Comment by mykola-mokhnach** [comment]
> The breakdown of the build matrix is:
> 
> All checks (unit, e2e) other than ios-simulator run against supported Node.js versions
> The ios-simulator unit tests run
> Assuming 2. completes successfully, ios-simulator's e2e tests will run against the default Node.js version (I think it's v14 atm)

I would prefer if we could breakdown tests by unit/e2e, so all unit tests are being executed in the first stage while e2e tests are executed in the second stage. I don't see a need to make simulator tests so unique, because other modules also require some special e2e setup and then it would be necessary to change the script again (e.g. the idb one requires facebook idb tool to be installed)

**Comment by mykola-mokhnach** [comment]
Also, it is hard to parse the tests results stuff in general if it is named common. It would be much easier to see what exatcly failed if these jobs would include corresponding component names

**Comment by boneskull** [comment]
@mykola-mokhnach 

> Also, it is hard to parse the tests results stuff in general if it is named common. It would be much easier to see what exatcly failed if these jobs would include corresponding component names

I agree that'd be nice (and was my original idea), but:

a) the status quo in `appium/appium` is this
b) it would _significantly_ increase execution time because it'd need _number-of-packages * 3_ (node.js versions) VMs to boot up, retrieve `node_modules` from cache and compile all packages

so I'm inclined to leave it

**Comment by boneskull** [comment]
> I would prefer if we could breakdown tests by unit/e2e, so all unit tests are being executed in the first stage while e2e tests are executed in the second stage. I don't see a need to make simulator tests so unique, because other modules also require some special e2e setup and then it would be necessary to change the script again (e.g. the idb one requires facebook idb tool to be installed)

Hmm, it must be that not all the E2E tests are running, then?  Or they are getting skipped?

**Comment by boneskull** [comment]
yeah they aren't getting run.

**Comment by mykola-mokhnach** [comment]
Unit tests are fast, so we could easily run all of them for all supported Node versions.
It would make sense to run e2e tests for only actually modified packages. And leave other packages for manual execution if needed

**Comment by boneskull** [comment]
> Unit tests are fast, so we could easily run all of them for all supported Node versions.

Yes, we are running unit tests for all supported node versions.  However, we're not running these as separate _jobs_ because the setup of the VM takes longer than the unit tests do.

> It would make sense to run e2e tests for only actually modified packages. And leave other packages for manual execution if needed

I did consider this, but I think it's something we can work on later--there are interdependencies between these packages, so we'd need to make sure _x_ is run if _y_ is modified.

**Review by jlipps — APPROVED** [review]
code changes look good. i don't have a strong opinion on the question of exactly how to divide up the tests, so i'll defer to you two.
